File: backtracking/1415_The_k-th_Lexicographical_String_of_All_Happy_Strings_of_Length_n/1415_The_k-th_Lexicographical_String_of_All_Happy_Strings_of_Length_n.py

class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        self.res = ""
        self.n = n
        self.k = k
        def dfs(cur, tmp):
            if len(cur) == self.n:
                self.k -= 1
                if self.k == 0:
                    self.res = cur
                    return
                return
            for i in tmp:
                if self.res:
                    return self.res
                new_tmp = [j for j in ['a', 'b', 'c'] if j != i]
                dfs(cur+i, new_tmp)
        dfs("", ['a','b', 'c'])
        return self.res


n = 1
k = 3
res = Solution().getHappyString(n, k)
print(res)

File: backtracking/1415_The_k-th_Lexicographical_String_of_All_Happy_Strings_of_Length_n/1415_The_k-th_Lexicographical_String_of_All_Happy_Strings_of_Length_n.go
package main

import "fmt"

func getHappyString(n int, k int) string {
	res := ""
	num := 0
	var dfs func(i int, buf []byte)
	dfs = func(i int, buf []byte) {
		if i == n {
			num++
			if num == k {
				res = string(buf)
			}
			return
		}
		for j := byte('a'); j < 'd'; j++ {
			if i == 0 || j != buf[i-1] {
				buf[i] = j
				dfs(i+1, buf)
			}
		}
	}
	dfs(0, make([]byte, n))
	return res
}

func main() {
	n := 1
	k := 3
	res := getHappyString(n, k)
	fmt.Println(res)
}
